README.md
RNAAgeCalc
===========

`RNAAgeCalc` is a multi-tissue transcriptionals age calculator. 

The `RNAAgeCalc` paper can be found [here](https://doi.org/10.1371/journal.pone.0237006).   
The Bioconductor package can be found [here](https://bioconductor.org/packages/release/bioc/html/RNAAgeCalc.html).    
The Bioconductor package vignette can be found [here](https://bioconductor.org/packages/release/bioc/vignettes/RNAAgeCalc/inst/doc/RNAAge-vignette.html).    
An interactive Shiny App can be found [here](https://xuren2120.shinyapps.io/RNAAgeCalcshiny/).    
The Shiny App demo can be found [here](http://www.ams.sunysb.edu/~pfkuan/RNAAgeCalc/instructions.html).    
The Python version of RNAAgeCalc can be found [here](https://pypi.org/project/racpy).

Package installation
------------
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))  install.packages("BiocManager")    
BiocManager::install("RNAAgeCalc")    
```

Citation
------------
Ren, X., & Kuan, P. F. (2020). RNAAgeCalc: A multi-tissue transcriptional age calculator. PloS one, 15(8), e0237006. 

@article{ren2020rnaagecalc,    
title={RNAAgeCalc: A multi-tissue transcriptional age calculator},    
author={Ren, Xu and Kuan, Pei-Fen},    
journal={PLOS ONE},    
volume={15},    
number={8},    
pages={1-21},    
year={2020},    
publisher={Public Library of Science}    
}


inst/NEWS
Changes in version 1.1.2 (2020-09-20)
+ Added citation and README

Changes in version 1.1.1 (2020-06-24)
+ Replaced "DESeq" with "DESeq2" throughout the package

Changes in version 0.99.1 (2019-08-30)
+ Package created


tests/testthat.R
library(testthat)
library(RNAAgeCalc)

test_check("RNAAgeCalc")


tests/testthat/test_main.R
context("Test pred_age, makeplot")

data(fpkmExample)
library(SummarizedExperiment)

test_that("Test invalid input", {
    exprdata1 = 1
    expect_error(predict_age(exprdata = exprdata1, exprtype = "FPKM"))
    
    exprdata2 = matrix(c("a", "b", "c", "d"), 2, 2)
    expect_error(predict_age(exprdata = exprdata2, exprtype = "FPKM"))
    
    exprdata3 = matrix(c(1:8, -1), 3, 3)
    rownames(exprdata3) = c("a", "b", "c")
    colnames(exprdata3) = c("sample1", "sample2", "sample3")
    expect_error(predict_age(exprdata = exprdata3, exprtype = "FPKM"))
    
    exprdata4 = data.frame(sample1 = c(1,2,3), sample2 = c("a", "b", "c"),
                           sample3 = as.factor(c("ctr", "ctr", "trt")))
    expect_error(predict_age(exprdata = exprdata4, exprtype = "FPKM"))
    
    exprdata5 = data.frame(matrix(c(1:8, -1), 3, 3))
    expect_error(predict_age(exprdata = exprdata5, exprtype = "FPKM"))
    
    exprdata6 = matrix(1:9, 3, 3)
    expect_error(predict_age(exprdata = exprdata6))
    rownames(exprdata6) = c("a", "b", "c")
    expect_error(predict_age(exprdata = exprdata6))
    colnames(exprdata6) = c("sample1", "sample2", "sample3")
    expect_error(predict_age(exprdata = exprdata6, exprtype = "unknown"))
    
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM", 
                             stype = "unknown"))
    
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM", 
                             maxp = "unknown"))
    
    se1 = 1
    expect_error(predict_age_fromse(se = se1, exprtype = "FPKM"))
    
    colData = data.frame(age = c(40, 50))
    se2 = SummarizedExperiment(assays=list(unknown=fpkm), colData=colData)
    expect_error(predict_age_fromse(se = se2, exprtype = "FPKM"))
    
    se3 = SummarizedExperiment(assays=list(FPKM=fpkm), colData=colData)
    expect_error(predict_age_fromse(se = se3, exprtype = "counts"))
    
    se4 = SummarizedExperiment(assays=list(counts=fpkm), colData=colData)
    expect_error(predict_age_fromse(se = se4, exprtype = "FPKM"))
    
    # These are some intermediate tests when we are writing the 
    # predict_age_fromse() function. It no longer holds once we finish writing
    # the function. Thus, they are comment.
    # rowData = DataFrame(bp_length = runif(nrow(fpkm)))
    # se5 = SummarizedExperiment(assays=list(counts=fpkm), colData=colData, 
    #                            rowData = rowData)
    # res = predict_age_fromse(se5, tissue = NULL, exprtype="counts")
    # expect_identical(res, rowData[["bp_length"]])
    
    # se5 = SummarizedExperiment(assays=list(fpkm=fpkm), colData=colData)
    # res = predict_age_fromse(se5, tissue = NULL)
    # expect_identical(res, data.frame(sampleid = colnames(fpkm), 
    #                                 age = c(40, 50)))
    
    se5 = SummarizedExperiment(assays=list(fpkm=fpkm), colData=colData)
    expect_error(predict_age_fromse(se = se5, exprtype = "unknown"))
    
    expect_error(predict_age_fromse(se = se5, exprtype = "FPKM", 
                                    stype = "unknown"))
    
    expect_error(predict_age_fromse(se = se5, exprtype = "FPKM", 
                                    maxp = "unknown"))
    
    res1 = predict_age(exprdata = fpkm, tissue = "brain", signature = "GTExAge", 
                       chronage = data.frame(sampleid = colnames(fpkm), 
                                             age = c(40, 50)))
    res2 = predict_age_fromse(se = se5, tissue = "brain", 
                              signature = "GTExAge")
    expect_identical(res1, res2)
    
    se6 = SummarizedExperiment(assays=list(fpkm=fpkm))
    res1 = predict_age(exprdata = fpkm, tissue = "brain", 
                       signature = "GTExAge")
    res2 = predict_age_fromse(se = se6, tissue = "brain", 
                              signature = "GTExAge")
    expect_identical(res1, res2)
})


test_that("Test tissue, signature, idtype, and chronage arguments", {
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM"), NA)
    expect_error(predict_age(exprdata = fpkm, tissue = "breaste",
        exprtype = "FPKM"), NA)
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM",
        idtype = "symbole"))
    
    colData = data.frame(age = c(40, 50))
    se1 = SummarizedExperiment(assays=list(fpkm=fpkm), colData=colData)
    expect_error(predict_age_fromse(se = se1, exprtype = "FPKM"), NA)
    expect_error(predict_age_fromse(se = se1, tissue = "breaste",
                                    exprtype = "FPKM"), NA)
    expect_error(predict_age_fromse(se = se1, exprtype = "FPKM",
                                    idtype = "symbole"))
    
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        signature = "unknown"))
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        signature = "DESeq2"),NA)
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        signature = "DESeq"),NA)
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        tissue = "brain"),NA)
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        tissue = "brain", signature = "unknown"))
    
    temp1 = predict_age(exprdata = fpkm, idtype = "SYMBOL",
        signature = "DESeq2")
    temp2 = predict_age(exprdata = fpkm, idtype = "SYMBOL",
        signature = "DESeq")
    expect_identical(temp1, temp2)
    
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
                                    signature = "unknown"))
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
                                    signature = "DESeq2"),NA)
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
                                    signature = "DESeq"),NA)
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
                                    tissue = "brain"),NA)
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL", 
                                    tissue = "brain", signature = "unknown"))
    
    temp1 = predict_age_fromse(se = se1, idtype = "SYMBOL", 
        signature = "DESeq2")
    temp2 = predict_age_fromse(se = se1, idtype = "SYMBOL", 
        signature = "DESeq")
    expect_identical(temp1, temp2)
    

    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        tissue = "unknown", signature = "unknown"))
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        tissue = "unknown", signature = "DESeq2"), NA)
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        tissue = "unknown", signature = "DESeq"), NA)
    
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
        tissue = "unknown", signature = "DESeq2"), NA)
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
        tissue = "unknown", signature = "DESeq"), NA)

    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        tissue = "breast"),NA)
    expect_error(predict_age(exprdata = fpkm, idtype = "SYMBOL",
        tissue = "unknown"),NA)
    
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
                             tissue = "breast"),NA)
    expect_error(predict_age_fromse(se = se1, idtype = "SYMBOL",
                             tissue = "unknown"),NA)

    ## test the chronage argument
    chronage = data.frame(sampleid = colnames(fpkm), age = c(30,50))
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM",
        chronage = chronage),NA)

    res = predict_age(exprdata = fpkm, exprtype = "FPKM")
    expect_error(makeplot(res))

    res = predict_age(exprdata = fpkm, exprtype = "FPKM", chronage = chronage)
    expect_warning(makeplot(res))

    #chronage1 = data.frame(sampleid = c(30,50), age = c(30,50))
    #expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM",
    #    chronage = chronage1))

    chronage2 = data.frame(sampleid = colnames(fpkm),
        age = c("unknown", "unknown"))
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM",
        chronage = chronage2))

    chronage3 = data.frame(sampleid = c("unknown1","unknown2"), age = c(30,50))
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM",
        chronage = chronage3))

    chronage4 = data.frame(sampleid = colnames(fpkm), age = c(NA,-3))
    expect_error(predict_age(exprdata = fpkm,
        exprtype = "FPKM", chronage = chronage4))

    chronage5 = data.frame(sampleid = colnames(fpkm),
                           age = c(30,50), age2 = c(30,50))
    expect_error(predict_age(exprdata = fpkm, exprtype = "FPKM",
        chronage = chronage5),NA)

    fpkm_large = cbind(fpkm, fpkm, fpkm, fpkm)
    fpkm_large = cbind(fpkm_large, fpkm_large, fpkm_large, fpkm_large)
    colnames(fpkm_large) = paste0("sample",1:32)
    chronage6 = data.frame(sampleid = colnames(fpkm_large), age = 1:32)
    expect_error(predict_age(exprdata = fpkm_large, exprtype = "FPKM",
        chronage = chronage6),NA)
    res = predict_age(exprdata = fpkm_large, exprtype = "FPKM",
                      chronage = chronage6)
    expect_error(makeplot(res), NA)
    
    chronage7 = rbind(chronage6, data.frame(sampleid = "sample2", age = 2))
    expect_error(predict_age(exprdata = fpkm_large, exprtype = "FPKM",
                             chronage = chronage7))

    ## check for imputation
    tissue = "brain"
    signature = "DESeq2"
    genes_required = names(genelist_all[[tissue]][[signature]])[-1]
    fpkm2 = fpkm
    fpkm2 = fpkm2[which(!rownames(fpkm2)%in%genes_required[c(1,10,20,30)]),]
    fpkm2 = fpkm2[which(rownames(fpkm2)%in%genes_required),]
    ## fpkm2 contains missing genes
    expect_error(predict_age(fpkm2, "brain"),NA)

    ## partial row NA
    fpkm3 = fpkm
    fpkm3 = fpkm3[which(rownames(fpkm3)%in%genes_required),]
    fpkm3[1,2] = NA
    fpkm3[3,1] = NA
    expect_error(predict_age(fpkm3, "brain"),NA)

    ## some other tests:
    ## These are some tests for our internal data, it has nothing to do
    ## with the user facing functions. Since some of the code takes too long,
    ## these tests are commented in order to pass the Bioconductor check
    ## time limit

    ## test whether the id type conversion work well
    #suppressMessages(library(recount))
    ## get GTEx brain data from recount package
    #load(url("http://duffel.rail.bio/recount/v2/SRP012682/rse_gene_brain.Rdata"))
    #rse = scale_counts(rse_gene)
    ## get RPKM matrix
    #fpkm <- getRPKM(rse)

    ## this is the gene map used in our paper
    #gene_map = read.csv("https://raw.githubusercontent.com/reese3928/RNAAgeData/master/gene_map_03182019.csv")

    #tempid = which(!duplicated(vapply(strsplit(rownames(fpkm),split="\\."), '[', 1, FUN.VALUE=character(1))))
    #fpkm = fpkm[tempid,]
    #rownames(fpkm) = vapply(strsplit(rownames(fpkm),split="\\."), '[', 1, FUN.VALUE=character(1))
    #fpkm2 = fpkm[sample(nrow(fpkm)),]

    #exprdata = fpkm2
    #idtype = "ENSEMBL"
    #mapid = suppressMessages(select(org.Hs.eg.db, rownames(exprdata),
    #                                columns = "SYMBOL", keytype = idtype))
    #mapid = mapid[which(!is.na(mapid$SYMBOL)),]
    #mapid = mapid[!duplicated(mapid[,idtype]),]
    #ind1 = which(!mapid$SYMBOL %in% mapid$SYMBOL[duplicated(mapid$SYMBOL)])
    #mapid = mapid[ind1,]
    #rownames(mapid) = as.character(mapid[,idtype])
    #exprdata = exprdata[rownames(exprdata)%in%rownames(mapid),]
    #genesymbol = mapid[rownames(exprdata),"SYMBOL"]
    #rownames(exprdata) = genesymbol
    #all(gene_map$SYMBOL%in%rownames(exprdata))

})










tests/testthat/test_count2FPKM.R
context("Test count2FPKM")

data(countExample)

test_that("Test count2FPKM", {
    rawcount2 = rawcount
    rownames(rawcount2)[c(1000,2000,3000)] = c("ABC1", "ABC2", "ABC3")
    expect_warning(count2FPKM(rawcount2, genelength = NULL))

    genelength1 = rep("unknown", nrow(rawcount))
    expect_error(count2FPKM(rawcount, genelength = genelength1))

    genelength2 = matrix(1:9,3,3)
    expect_error(count2FPKM(rawcount, genelength = genelength2))

    genelength3 = rep(1, nrow(rawcount)-1)
    expect_error(count2FPKM(rawcount, genelength = genelength3))

    rawcount3 = rawcount[1:5,]
    rownames(rawcount3) = c("ENSG00000000003", "ENSG00000000004",
        "ENSG00000000005", "ENSG00000000419", "ENSG00000000457")
    expect_warning(count2FPKM(rawcount3, idtype = "ENSEMBL"))

})


R/makeplot.R


#' @title Make plot to visualize RNA age
#'
#' @description This function makes plots to visualize the relationship
#' between chronological age and RNA age.
#' @param res a data frame returned by `predict_age` function. If the
#' chronological age is not provided when using `predict_age` function,
#' visulization cannot be made.
#' @param main title of the plot
#' @param xlab label of x-axis
#' @param ylab label of y-axis
#' @importFrom ggplot2 ggplot aes geom_point geom_smooth ggtitle xlab ylab
#' theme_bw theme element_text
#' @export
#' @return the plot which shows RNA age vs chronological age
#' @examples
#' data(fpkmExample)
#' fpkm_large = cbind(fpkm, fpkm, fpkm, fpkm)
#' fpkm_large = cbind(fpkm_large, fpkm_large, fpkm_large, fpkm_large)
#' colnames(fpkm_large) = paste0("sample",1:32)
#' chronage = data.frame(sampleid = colnames(fpkm_large), age = 1:32)
#' res = predict_age(exprdata = fpkm_large, exprtype = "FPKM",
#' chronage = chronage)
#' makeplot(res)

makeplot = function(res, main = "RNA age vs chronological age",
                    xlab = "chronological age", ylab = "RNA Age"){
    if(!"ChronAge"%in%colnames(res)){
        stop("Chronological age is not found in res data frame.")
    }
    if(nrow(na.omit(res))<30){
        warning("Less than 30 samples. The linear regression on RNA age vs
chronological age may not be reliable.")
    }
    ggplot(res, aes(x=ChronAge, y=RNAAge)) + geom_point(shape=19,size=1) +
        geom_smooth(method=lm) + ggtitle(main) + xlab(xlab) + ylab(ylab) +
        theme_bw() + theme(plot.title = element_text(hjust = 0.5))
}




R/count2FPKM.R

#' @title Converting gene expression data from raw count to FPKM
#' @description This function converts gene expression data from raw count to
#' FPKM by using \code{\link[recount]{getRPKM}} 
#'
#' @param rawcount a matrix or data frame which contains gene expression counts
#' data.
#' @param genelength gene length in bp. The size of `genelength` should be
#' equal to the number of rows in `rawcount`. This argument is optional. If
#' not provided, gene length is obtained from the internal database.
#' @param idtype a string which indicates the gene id type in rawcount matrix.
#' It should be one of "SYMBOL", "ENSEMBL", "ENTREZID" or "REFSEQ".
#' Default is "SYMBOL".
#' @export
#' @import org.Hs.eg.db
#' @importFrom AnnotationDbi select
#' @importFrom recount getRPKM
#' @importFrom SummarizedExperiment SummarizedExperiment
#' @return a data frame contains FPKM.
#' @references Carlson M (2019). org.Hs.eg.db: Genome wide annotation for
#' Human. R package version 3.8.2.
#' @references Collado-Torres, Leonardo, et al. "Reproducible RNA-seq analysis 
#' using recount2." Nature biotechnology 35.4 (2017): 319-321.
#' @examples
#' data(countExample)
#' head(rawcount)
#' fpkm = count2FPKM(rawcount)
#' head(fpkm)

count2FPKM = function(rawcount, genelength = NULL, idtype = "SYMBOL"){
    if(is.null(genelength)){
        if(idtype!="ENSEMBL"){
            ## convert the gene id to ENSEMBL to match the gene id in gene
            ## length database
            temp = suppressMessages(select(org.Hs.eg.db, rownames(rawcount),
                columns = "ENSEMBL", keytype = idtype))
            outcome = temp[temp$ENSEMBL%in%rownames(lengthDB),]
            outcome = outcome[!duplicated(outcome[,idtype]),]
            ## ind is all the one to one map
            ind = which(!outcome$ENSEMBL %in%
                outcome$ENSEMBL[duplicated(outcome$ENSEMBL)])
            outcome1 = outcome[ind,]   ## outcome1 is all the 1-1 map

            matchid = match(rownames(rawcount),outcome1[,idtype])
            genelength = lengthDB[outcome1$ENSEMBL[matchid],"bp_length"]
        }else{
            genelength = lengthDB[rownames(rawcount), "bp_length"]
        }
        
    }else{
        stopifnot(is.vector(genelength))
        stopifnot(is.numeric(genelength))
        if(any(genelength<0, na.rm = TRUE)){
            stop("genelength cannot contain negative value(s).")
        }
        if(length(genelength)!=nrow(rawcount)){
            stop("The size of genelength vector should be equal to the number 
of rows in rawcount.")
        }
    }
    
    if(any(is.na(genelength))){
        warning("Can't find gene length for ",
                round(sum(is.na(genelength))/nrow(rawcount)*100,4),
                "% genes when converting raw count to FPKM.")
    }

    rowData = data.frame(bp_length = genelength)
    se = SummarizedExperiment(assays=list(counts=rawcount), rowData=rowData)
    df = data.frame(getRPKM(se))
    df
}


R/exampledata.R
#' @title An example of FPKM data
#'
#' @description This is an example of FPKM data. It is the gene expression 
#' FPKM data from The genotype-tissue expression (GTEx) project brain samples. 
#' The data was downloaded from 
#' recount2(https://jhubiostatistics.shinyapps.io/recount/). For 
#' illustration purpose, only two samples were included. The R script to 
#' generate this example data can be found in inst/scripts/createExamples.R.
#' @docType data
#' @name fpkm
#' @usage fpkm
#' @format A data frame which contains the fpkm data. The dimension of this 
#' data frame is 24989 by 2. Each row is a gene and each column is a sample. 
#' @keywords datasets
#' @references Collado-Torres, Leonardo, et al. "Reproducible RNA-seq analysis 
#' using recount2." Nature biotechnology 35.4 (2017): 319-321.
#' @references Lonsdale, John, et al. "The genotype-tissue expression (GTEx) 
#' project." Nature genetics 45.6 (2013): 580.
NULL

#' @title An example of RNASeq counts data
#'
#' @description An example of RNASeq counts data. It is the gene expression 
#' counts data from The genotype-tissue expression (GTEx) project brain samples. 
#' The data was downloaded from 
#' recount2(https://jhubiostatistics.shinyapps.io/recount/). For 
#' illustration purpose, only two samples were included. The R script to 
#' generate this example data can be found in inst/scripts/createExamples.R.
#' @docType data
#' @name rawcount
#' @usage rawcount
#' @format A data frame which contains the RNASeq counts data. The dimension of 
#' this data frame is 24989 by 2. Each row is a gene and each column is a 
#' sample. 
#' @keywords datasets
#' @references Collado-Torres, Leonardo, et al. "Reproducible RNA-seq analysis 
#' using recount2." Nature biotechnology 35.4 (2017): 319-321.
#' @references Lonsdale, John, et al. "The genotype-tissue expression (GTEx) 
#' project." Nature genetics 45.6 (2013): 580.
NULL

data-raw/internalData.R


## This script generates all the internal data used in the package.
suppressMessages(library(AnnotationDbi))
library(org.Hs.eg.db)
suppressMessages(library(recount))

## get GTEx brain data from recount package
load(url("http://duffel.rail.bio/recount/v2/SRP012682/rse_gene_brain.Rdata"))
lengthDB = rowData(rse_gene)
tempid = which(!duplicated(vapply(strsplit(rownames(lengthDB),split="\\."),
                                  '[', 1, FUN.VALUE=character(1))))
lengthDB = lengthDB[tempid,]
rownames(lengthDB) = vapply(strsplit(rownames(lengthDB),split="\\."), '[', 1,
                            FUN.VALUE=character(1))
lengthDB = lengthDB[,setdiff(colnames(lengthDB), c("gene_id","symbol")),
                    drop = FALSE]
lengthDB = data.frame(lengthDB)


library(glmnet)
library(devtools)
tissues = c("adipose_tissue", "adrenal_gland", "blood", "blood_vessel", "brain",
            "breast", "colon", "esophagus", "heart", "liver", "lung", "muscle",
            "nerve", "ovary", "pancreas", "pituitary", "prostate",
            "salivary_gland", "skin", "small_intestine", "spleen", "stomach",
            "testis", "thyroid", "uterus", "vagina")

signatures = c("DESeq", "Pearson", "Dev", "deMagalhaes", "GenAge",
               "GTExAge", "Peters", "all")

genelist_one = vector("list", length(signatures))
names(genelist_one) = signatures
genelist_all = vector("list", length(tissues))
names(genelist_all) = tissues
for(i in 1:length(genelist_all)) genelist_all[[i]] = genelist_one
genelist_cau = genelist_all

dir = "/Users/renxu/Desktop/Rpackage/RNAAgeData/all_samples"
for(t in tissues){
    message("tissue: ", t)
    for(sig in signatures){
        load(file.path(dir, t, paste0(sig,".RData")))
        coefs = coef(opt_model, s = "lambda.min")[,1]
        genelist_all[[t]][[sig]] = coefs[which(coefs!=0)]
    }
}

for(sig in signatures[-1]){
    load(file.path(dir, "all_tissues", paste0(sig,".RData")))
    coefs = coef(opt_model, s = "lambda.min")[,1]
    genelist_all[["all_tissues"]][[sig]] = coefs[which(coefs!=0)]
}

for(t in tissues) names(genelist_all[[t]])[1] = "DESeq2"

dir2 = "/Users/renxu/Desktop/Rpackage/RNAAgeData/Caucasian"
for(t in tissues){
    message("tissue: ", t)
    for(sig in signatures){
        load(file.path(dir2, t, paste0(sig,".RData")))
        coefs = coef(opt_model, s = "lambda.min")[,1]
        genelist_cau[[t]][[sig]] = coefs[which(coefs!=0)]
    }
}

for(sig in signatures[-1]){
    load(file.path(dir2, "all_tissues", paste0(sig,".RData")))
    coefs = coef(opt_model, s = "lambda.min")[,1]
    genelist_cau[["all_tissues"]][[sig]] = coefs[which(coefs!=0)]
}

for(t in tissues) names(genelist_cau[[t]])[1] = "DESeq2"

use_data(lengthDB, genelist_all, genelist_cau, internal = TRUE, 
         compress = "xz", overwrite = TRUE)


## some tests
load(file.path(dir, "breast", "DESeq.RData"))
load("data/FPKMExample.RData")
geneid = rownames(coef(opt_model, s = "lambda.min"))[-1]
fpkm1 = fpkm[geneid,]
dat.test = t(log2(fpkm1+1))
predict(opt_model, newx = dat.test, s = "lambda.min")

geneid2 = names(genelist_all[["breast"]][["DESeq"]])[-1]
fpkm2 = fpkm[geneid2,]
dat.test2 = t(log2(fpkm2+1))
coefs = genelist_all[["breast"]][["DESeq"]]
apply(dat.test2, 1, function(x) sum(x*coefs[-1])+coefs[1])






inst/scripts/createExamples.R


## This script generates example raw count, FPKM data
suppressMessages(library(recount))
## get GTEx brain data from recount package
load(url("http://duffel.rail.bio/recount/v2/SRP012682/rse_gene_brain.Rdata"))
rse = scale_counts(rse_gene)
## get RPKM matrix
fpkm <- getRPKM(rse)

## this is the gene map used in our paper
gene_map = read.csv("https://raw.githubusercontent.com/reese3928/RNAAgeData/master/gene_map_03182019.csv")

tempid = which(!duplicated(vapply(strsplit(rownames(fpkm),split="\\."), '[', 1, FUN.VALUE=character(1))))
fpkm = fpkm[tempid,]
rownames(fpkm) = vapply(strsplit(rownames(fpkm),split="\\."), '[', 1, FUN.VALUE=character(1))
fpkm = fpkm[as.character(gene_map$ENSEMBL),]
rownames(fpkm) = as.character(gene_map$SYMBOL)
fpkm = fpkm[,1:2]  ## subset the samples to make the example size smaller
save(fpkm, file = "data/fpkmExample.RData", compress = "xz")


rawcount = assay(rse)
tempid = which(!duplicated(vapply(strsplit(rownames(rawcount),split="\\."), '[', 1, FUN.VALUE=character(1))))
rawcount = rawcount[tempid,]
rownames(rawcount) = vapply(strsplit(rownames(rawcount),split="\\."), '[', 1, FUN.VALUE=character(1))
rawcount = rawcount[as.character(gene_map$ENSEMBL),]
rownames(rawcount) = as.character(gene_map$SYMBOL)
rawcount = rawcount[,1:2]  ## subset the samples to make the example size smaller
save(rawcount, file = "data/countExample.RData", compress = "xz")






R/predict_age_fromse.R

#' @title Calculate RNA age using SummarizedExperiment
#' @description This function takes SummarizedExperiment object as input and 
#' calculates RNA age. 
#' @param se a \code{\link{SummarizedExperiment}} object. The assays(se) should
#' contain gene expression data. The name of assays(se) should be either "FPKM"
#' or "counts". Use `exprtype` argument to specify the type of gene expression 
#' data provided. Users are able to provide the chronological age of samples
#' using colData(se). This is optional. If provided, the column name for 
#' chronological age in colData(se) should be "age". If some samples' 
#' chronological age are not available, users are expected to set the 
#' chronological age in colData(se) to NA. If chronological age is not 
#' provided, the age acceleration residual will not be calculated. See package 
#' vignette for the definition of age acceleration residual. In addition, users 
#' are able to provide their own gene length using rowData(se). This is also
#' optional. If using `exprtype = "FPKM"`, the provided gene length will be 
#' ignored. If provided, the column name for gene length in rowData(se) should 
#' be "bp_length". The function will convert raw count to FPKM by the
#' user-supplied gene length. Otherwise, gene length is obtained from the
#' internal database. See below for an example of se object.
#' @param tissue a string indicate which tissue the gene expression data is
#' obtained from. Users are expected to provide one of the following tissues.
#' If the tissue argument is not provide or the provided tissue is not in this
#' list, then the age predictor trained on all tissues will be used to
#' calculate RNA age.
#' \itemize{
#'   \item adipose_tissue
#'   \item adrenal_gland
#'   \item blood
#'   \item blood_vessel
#'   \item brain
#'   \item breast
#'   \item colon
#'   \item esophagus
#'   \item heart
#'   \item liver
#'   \item lung
#'   \item muscle
#'   \item nerve
#'   \item ovary
#'   \item pancreas
#'   \item pituitary
#'   \item prostate
#'   \item salivary_gland
#'   \item skin
#'   \item small_intestine
#'   \item spleen
#'   \item stomach
#'   \item testis
#'   \item thyroid
#'   \item uterus
#'   \item vagina
#' }
#' @param exprtype either "counts" or "FPKM". For RPKM data, please use
#' `exprtype` = "FPKM".
#' @param idtype a string which indicates the gene id type in `exprdata`.
#' It should be one of "SYMBOL", "ENSEMBL", "ENTREZID" or "REFSEQ". Default is
#' "SYMBOL".
#' @param stype a string which specifies which version of pre-trained 
#' calculators to be used. It should be either "all" or "Caucasian". "all" 
#' means samples from all races (American Indian/Alaska Native, Asian, 
#' Black/African American, and Caucasian) are used to obtain the 
#' pre-trained calculator. "Caucasian" means only the Caucasian samples are 
#' used to build up the pre-trained calculator.
#' @param signature a string which indicates the age signature to use when
#' calculating RNA age. This argument is not required. \cr
#' In the case that this argument is not provided, if `tissue` argument is also
#' provided and the tissue is in the list above, the tissue specific age
#' signature given by our DESeq2 analysis result on GTEx data will be used.
#' Otherwise, the across tissue signature "GTExAge" will be used. \cr
#' In the case that this argument is provided, it should be one of the
#' following signatures. A detailed description of the meaning of these
#' signatures is given in the package vignette.
#' \itemize{
#'   \item DESeq2
#'   \item Pearson
#'   \item Dev
#'   \item deMagalhaes
#'   \item GenAge
#'   \item GTExAge
#'   \item Peters
#'   \item all
#' }
#' @param maxp the maxp argument used in \code{\link[impute]{impute.knn}} 
#' function. This is optional. 
#' @return a data frame contains RNA age.
#' @importFrom SummarizedExperiment assayNames assay colData rowData
#' @importFrom methods is
#' @export
#' @examples 
#' library(SummarizedExperiment)
#' data(fpkmExample)
#' colData = data.frame(age = c(40, 50))
#' se = SummarizedExperiment(assays=list(FPKM=fpkm), 
#' colData=colData)
#' res = predict_age_fromse(se = se, exprtype = "FPKM")

predict_age_fromse <- function(se, tissue, exprtype = c("FPKM", "counts"), 
    idtype = c("SYMBOL", "ENSEMBL", "ENTREZID", "REFSEQ"),
    stype = c("all", "caucasian"), signature = NULL, maxp = NULL){
    
    if (!is(se, "SummarizedExperiment")) {
        stop("se should be a SummarizedExperiment object. ")
    }
    
    exprtype = match.arg(exprtype)
    ind = which(toupper(assayNames(se))==toupper(exprtype))
    if(length(ind)==0){
        stop("The assay name should contain ", exprtype)
    }
    
    exprdata = assay(se, ind)
    
    ## process gene length
    genelength = NULL
    if(exprtype=="counts"){
        lengthind = which(names(rowData(se))=="bp_length")
        if(lengthind > 0){
            genelength = rowData(se)[[lengthind]]
        }
    }
    
    ## process chronological age
    chronage = NULL
    ageind = which(toupper(names(colData(se)))=="AGE")
    if(length(ageind)>0){
        chronage = data.frame(sampleid = rownames(colData(se)), 
                              age = colData(se)[[ageind]])
    }
    
    if(missing(tissue)){
        res = predict_age(exprdata, exprtype = exprtype, idtype = idtype,
                          stype = stype, signature = signature, 
                          genelength = genelength, chronage = chronage, 
                          maxp = maxp)
    }else{
        res = predict_age(exprdata, tissue = tissue, 
                          exprtype = exprtype, idtype = idtype,
                          stype = stype, signature = signature, 
                          genelength = genelength, chronage = chronage,
                          maxp = maxp)
    }
    
    res
}





